engine/pkg/externalresource/manager/gc_coordinator.go

```diff
@@ -175,68 +172,59 @@ func (c *DefaultGCCoordinator) gcByStatusSnapshots(
 	}
 
 	var (
-		toGC []engineModel.JobID
-
-		// toRemove is used to remove meta records when
-		// the associated executors are offline.
-		// TODO adjust the mechanism when we implement S3 support.
-		toRemove []model.ExecutorID
+		toGCJobs      []engineModel.JobID
+		toGCExecutors []model.ExecutorID
 	)
 	for _, resMeta := range resources {
 		if _, exists := jobSnapshot[resMeta.Job]; !exists {
 			// The resource belongs to a deleted job.
-			toGC = append(toGC, resMeta.Job)
+			toGCJobs = append(toGCJobs, resMeta.Job)
 			continue
 		}
 
 		if _, exists := executorSet[resMeta.Executor]; !exists {
 			// The resource belongs to an offlined executor.
-			toRemove = append(toRemove, resMeta.Executor)
+			tp, resName, err := resModel.PasreResourceID(resMeta.ID)
+			if err != nil {
+				return err
+			}
+			if tp == resModel.ResourceTypeLocalFile ||
+				tp == resModel.ResourceTypeS3 && resName == s3.DummyResourceName {
```

Comment:
I'm not clear about why we need a dummy resource for s3 resources. What if we remove the `resName == s3.DummyResourceName`?

Reply:
If S3 is configured, a dummy resource will be persisted by broker to indicate that its temporary files have not been cleaned, which is useful to prevent resource leaks.

Normally a broker will attempt to clean up temporary files and dummy resources before exiting. If this step fails, the dummy record is stored in Meta, which will be cleaned up by GCCoordinator eventually.

Hence, we need the dummy resource's meta to trigger gc of the offlined executor.

Reply:
Do other s3 normal resources store in meta too?

Reply:
Yes, only persistent resources are stored in meta.